File: copymove/src/main/scala/farjs/copymove/CopyMoveUi.scala
package farjs.copymove

import farjs.copymove.CopyMoveUi._
import farjs.copymove.CopyMoveUiAction._
import farjs.filelist.FileListActions._
import farjs.filelist._
import farjs.filelist.api.{FileListDir, FileListItem}
import farjs.filelist.history.{History, HistoryKind, HistoryProvider}
import farjs.ui.popup._
import farjs.ui.task.{Task, TaskAction}
import farjs.ui.theme.Theme
import scommons.nodejs.path
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.scalajs.js
import scala.util.Success

object CopyMoveUi {

  val copyItemsHistoryKind: HistoryKind = HistoryKind("farjs.copyItems", 50)

  private[copymove] var copyItemsStats: UiComponent[CopyItemsStatsProps] = CopyItemsStats
  private[copymove] var copyItemsPopup: UiComponent[CopyItemsPopupProps] = CopyItemsPopup
  private[copymove] var copyProcessComp: UiComponent[CopyProcessProps] = CopyProcess
  private[copymove] var messageBoxComp: ReactClass = MessageBox
  private[copymove] var moveProcessComp: UiComponent[MoveProcessProps] = MoveProcess
}

class CopyMoveUi(show: CopyMoveUiAction,
                 from: FileListData,
                 maybeTo: Option[FileListData]) extends FunctionComponent[FileListPluginUiProps] {

  protected def render(compProps: Props): ReactElement = {
    val historyProvider = HistoryProvider.useHistoryProvider()
    val (maybeTotal, setTotal) = useState[Option[Double]](None)
    val (maybeToPath, setToPath) = useState[Option[(String, String)]](None)
    val (inplace, setInplace) = useState(false)
    val (move, setMove) = useState(false)
    val (showPopup, setShowPopup) = useState(true)
    val (showStats, setShowStats) = useState(false)
    val (showMove, setShowMove) = useState(false)
    val copied = useRef(Set.empty[String])
    val currTheme = Theme.useTheme()

    val props = compProps.plain
    
    val onTopItem: js.Function1[FileListItem, Unit] = item => copied.current += item.name

    def onDone(path: String, toPath: String): () => Unit = { () =>
      val currSelected = from.state.selectedNames.toSet
      val updatedSelection = currSelected -- copied.current
      if (updatedSelection != currSelected) {
        from.dispatch(FileListParamsChangedAction(
          offset = from.state.offset,
          index = from.state.index,
          selectedNames = js.Set[String](updatedSelection.toSeq: _*)
        ))
      }

      val isInplace = inplace
      props.onClose()

      for {
        copyItemsHistory <- historyProvider.get(copyItemsHistoryKind).toFuture
        _ <- copyItemsHistory.save(History(path, js.undefined)).toFuture
      } yield ()

      val updateAction = from.actions.updateDir(from.dispatch, from.state.currDir.path)
      from.dispatch(updateAction)
      updateAction.task.result.toFuture.andThen {
        case Success(updatedDir) =>
          if (isInplace) from.dispatch(FileListItemCreatedAction(toPath, updatedDir.asInstanceOf[FileListDir]))
          else maybeTo.foreach(to => to.dispatch(to.actions.updateDir(to.dispatch, to.state.currDir.path)))
      }
    }

    def isMove: Boolean =
      move ||
        show == ShowMoveToTarget ||
        show == ShowMoveInplace

    def isInplace: Boolean = {
      inplace ||
        show == ShowCopyInplace ||
        show == ShowMoveInplace ||
        maybeTo.forall(_.state.currDir.path == from.state.currDir.path) && FileListState.selectedItems(from.state).isEmpty
    }

    val fromSelected = FileListState.selectedItems(from.state).toList
    val items =
      if (!isInplace && fromSelected.nonEmpty) fromSelected
      else FileListState.currentItem(from.state).toList

    def resolveTargetDir(move: Boolean, path: String): Future[(String, Boolean)] = {
      val dirF = for {
        dir <- from.actions.api.readDir(from.state.currDir.path, path).toFuture
        sameDrive <-
          if (move) checkSameDrive(from, dir.path)
          else Future.successful(false)
      } yield {
        (dir.path, sameDrive)
      }

      from.dispatch(TaskAction(Task("Resolving target dir", dirF)))
      dirF
    }

    val onAction: js.Function1[String, Unit] = { path =>
      val move = isMove
      val inplace = isInplace
      val resolveF =
        if (!from.actions.api.isLocal) Future.successful((path, false))
        else if (!inplace) resolveTargetDir(move, path)
        else Future.successful((path, true))

      resolveF.map { case (toPath, sameDrive) =>
        setInplace(inplace)
        setMove(move)
        setShowPopup(false)

        setToPath(Some(path -> toPath))
        if (move && sameDrive) setShowMove(true)
        else setShowStats(true)
      }
    }

    val maybeError = maybeToPath.filter(_ => !inplace).flatMap { case (_, toPath) =>
      val op = if (move) "move" else "copy"
      if (from.state.currDir.path == toPath) Some {
        s"Cannot $op the item\n${items.head.name}\nonto itself"
      }
      else if (toPath.startsWith(from.state.currDir.path + path.sep)) {
        val toSuffix = toPath.stripPrefix(from.state.currDir.path + path.sep)
        val maybeSelf = items.find(i => toSuffix == i.name || toSuffix.startsWith(i.name + path.sep))
        maybeSelf.map { self =>
          s"Cannot $op the item\n${self.name}\ninto itself"
        }
      }
      else None
    }

    <.>()(
      if (showPopup) Some {
        <(copyItemsPopup())(^.plain := CopyItemsPopupProps(
          move = isMove,
          path =
            if (!isInplace) maybeTo.map(_.state.currDir.path).getOrElse("")
            else FileListState.currentItem(from.state).map(_.name).getOrElse(""),
          items = js.Array(items: _*),
          onAction = onAction,
          onCancel = props.onClose
        ))()
      }
      else if (maybeError.isDefined) maybeError.map { error =>
        <(messageBoxComp)(^.plain := MessageBoxProps(
          title = "Error",
          message = error,
          actions = js.Array(MessageBoxAction.OK(props.onClose)),
          style = currTheme.popup.error
        ))()
      }
      else if (showStats) Some {
        <(copyItemsStats())(^.plain := CopyItemsStatsProps(
          dispatch = from.dispatch,
          actions = from.actions,
          fromPath = from.state.currDir.path,
          items = js.Array(items: _*),
          title = if (move) "Move" else "Copy",
          onDone = { total =>
            setTotal(Some(total))
            setShowStats(false)
          },
          onCancel = props.onClose
        ))()
      }
      else if (showMove) maybeToPath.map { case (path, toPath) =>
        <(moveProcessComp())(^.wrapped := MoveProcessProps(
          dispatch = from.dispatch,
          actions = from.actions,
          fromPath = from.state.currDir.path,
          items =
            if (!inplace) items.map(i => (i, i.name))
            else items.map(i => (i, toPath)),
          toPath =
            if (!inplace) toPath
            else from.state.currDir.path,
          onTopItem = onTopItem,
          onDone = onDone(path, toPath)
        ))()
      }
      else {
        for {
          (path, toPath) <- maybeToPath
          total <- maybeTotal
        } yield {
          <(copyProcessComp())(^.plain := CopyProcessProps(
            from = from,
            to =
              if (!inplace) maybeTo.getOrElse(from)
              else from,
            move = move,
            fromPath = from.state.currDir.path,
            items = js.Array((
              if (!inplace) items.map(i => CopyProcessItem(i, i.name))
              else items.map(i => CopyProcessItem(i, toPath))
            ): _*),
            toPath =
              if (!inplace) toPath
              else from.state.currDir.path,
            total = total,
            onTopItem = onTopItem,
            onDone = onDone(path, toPath)
          ))()
        }
      }
    )
  }
  
  private def checkSameDrive(from: FileListData, toPath: String): Future[Boolean] = {
    for {
      maybeFromRoot <- from.actions.api.getDriveRoot(from.state.currDir.path).toFuture
      maybeToRoot <- from.actions.api.getDriveRoot(toPath).toFuture
    } yield {
      val maybeSameDrive = for {
        fromRoot <- maybeFromRoot
        toRoot <- maybeToRoot
      } yield {
        fromRoot == toRoot
      }

      maybeSameDrive.getOrElse(false)
    }
  }
}

File: copymove/CopyMoveUi.mjs
/**
 * @import { FileListItem } from "@farjs/filelist/api/FileListItem.mjs"
 * @import { FileListAction } from "@farjs/filelist/FileListActions.mjs"
 * @import { FileListData } from "@farjs/filelist/FileListData.mjs"
 * @import { FileListPluginUiProps } from "@farjs/filelist/FileListPlugin.mjs"
 */
import path from "path";
import React, { useRef, useState } from "react";
import MessageBox from "@farjs/ui/popup/MessageBox.mjs";
import MessageBoxAction from "@farjs/ui/popup/MessageBoxAction.mjs";
import Task from "@farjs/ui/task/Task.mjs";
import TaskAction from "@farjs/ui/task/TaskAction.mjs";
import Theme from "@farjs/ui/theme/Theme.mjs";
import HistoryProvider from "@farjs/filelist/history/HistoryProvider.mjs";
import FileListState from "@farjs/filelist/FileListState.mjs";
import { isEqualSets, stripPrefix } from "@farjs/filelist/utils.mjs";
import CopyItemsPopup from "./CopyItemsPopup.mjs";
import CopyItemsStats from "./CopyItemsStats.mjs";
import CopyProcess from "./CopyProcess.mjs";
import MoveProcess from "./MoveProcess.mjs";

const h = React.createElement;

/**
 * @typedef {"ShowCopyToTarget"
 *  | "ShowCopyInplace"
 *  | "ShowMoveToTarget"
 *  | "ShowMoveInplace"
 * } CopyMoveUiAction
 */

/**
 * @typedef {{
 *  readonly show: CopyMoveUiAction;
 *  readonly from: FileListData;
 *  readonly maybeTo?: FileListData;
 * }} CopyMoveUiOptions
 */

/**
 * @param {CopyMoveUiOptions} options
 */
function CopyMoveUi({ show, from, maybeTo }) {
  /**
   * @param {FileListPluginUiProps} props
   */
  const CopyMoveUiComp = (props) => {
    const {
      copyItemsPopup,
      copyItemsStats,
      messageBoxComp,
      moveProcessComp,
      copyProcessComp,
    } = CopyMoveUi;

    const historyProvider = HistoryProvider.useHistoryProvider();
    const [maybeTotal, setTotal] = useState(
      /** @type {number | undefined} */ (undefined),
    );
    const [maybeToPath, setToPath] = useState(
      /** @type {[string, string] | undefined} */ (undefined),
    );
    const [inplace, setInplace] = useState(false);
    const [move, setMove] = useState(false);
    const [showPopup, setShowPopup] = useState(true);
    const [showStats, setShowStats] = useState(false);
    const [showMove, setShowMove] = useState(false);
    /** @type {React.MutableRefObject<Set<string>>} */
    const copied = useRef(new Set([]));
    const currTheme = Theme.useTheme();

    /** @type {(item: FileListItem) => any} */
    const onTopItem = (item) => copied.current.add(item.name);

    /** @type {(path: string, toPath: string) => () => void} */
    const onDone = (path, toPath) => () => {
      const currSelected = from.state.selectedNames;
      const newSelected = new Set([...currSelected]);
      copied.current.forEach((_) => newSelected.delete(_));
      if (!isEqualSets(currSelected, newSelected)) {
        /** @type {FileListAction} */
        const action = {
          action: "FileListParamsChangedAction",
          offset: from.state.offset,
          index: from.state.index,
          selectedNames: newSelected,
        };
        from.dispatch(action);
      }

      const isInplace = inplace;
      props.onClose();

      historyProvider
        .get(CopyItemsPopup.copyItemsHistoryKind)
        .then((copyItemsHistory) => copyItemsHistory.save({ item: path }));

      const updateAction = from.actions.updateDir(
        from.dispatch,
        from.state.currDir.path,
      );
      from.dispatch(updateAction);
      updateAction.task.result.then((updatedDir) => {
        if (isInplace) {
          /** @type {FileListAction} */
          const action = {
            action: "FileListItemCreatedAction",
            name: toPath,
            currDir: updatedDir,
          };
          from.dispatch(action);
        } else if (maybeTo !== undefined) {
          maybeTo.dispatch(
            maybeTo.actions.updateDir(
              maybeTo.dispatch,
              maybeTo.state.currDir.path,
            ),
          );
        }
      });
    };

    /** @type {() => boolean} */
    const isMove = () =>
      move || show === "ShowMoveToTarget" || show === "ShowMoveInplace";

    /** @type {() => boolean} */
    const isInplace = () =>
      inplace ||
      show === "ShowCopyInplace" ||
      show === "ShowMoveInplace" ||
      ((maybeTo === undefined ||
        maybeTo.state.currDir.path === from.state.currDir.path) &&
        FileListState.selectedItems(from.state).length === 0);

    const fromSelected = FileListState.selectedItems(from.state);
    const items =
      !isInplace() && fromSelected.length > 0
        ? fromSelected
        : [FileListState.currentItem(from.state)].filter(
            (_) => _ !== undefined,
          );

    /**
     * @param {boolean} move
     * @param {string} path
     * @returns {Promise<[string, boolean]>}
     */
    function resolveTargetDir(move, path) {
      /** @type {Promise<[string, boolean]>} */
      const dirP = (async () => {
        const dir = await from.actions.api.readDir(
          from.state.currDir.path,
          path,
        );
        const sameDrive = move ? await checkSameDrive(from, dir.path) : false;
        return [dir.path, sameDrive];
      })();

      from.dispatch(TaskAction(Task("Resolving target dir", dirP)));
      return dirP;
    }

    /** @type {(path: string) => void} */
    function onAction(path) {
      const move = isMove();
      const inplace = isInplace();

      /** @type {Promise<[string, boolean]>} */
      const resolveP = !from.actions.api.isLocal
        ? Promise.resolve([path, false])
        : !inplace
          ? resolveTargetDir(move, path)
          : Promise.resolve([path, true]);

      resolveP.then(([toPath, sameDrive]) => {
        setInplace(inplace);
        setMove(move);
        setShowPopup(false);

        setToPath([path, toPath]);
        if (move && sameDrive) setShowMove(true);
        else setShowStats(true);
      });
    }

    const maybeError = (() => {
      if (!maybeToPath || inplace) {
        return undefined;
      }
      const [_, toPath] = maybeToPath;
      const op = move ? "move" : "copy";
      if (from.state.currDir.path === toPath) {
        return `Cannot ${op} the item\n${items[0].name}\nonto itself`;
      }
      if (toPath.startsWith(from.state.currDir.path + path.sep)) {
        const toSuffix = stripPrefix(
          toPath,
          from.state.currDir.path + path.sep,
        );
        const self = items.find(
          (i) => toSuffix === i.name || toSuffix.startsWith(i.name + path.sep),
        );
        return self && `Cannot ${op} the item\n${self.name}\ninto itself`;
      }
      return undefined;
    })();

    return showPopup
      ? h(copyItemsPopup, {
          move: isMove(),
          path: (() => {
            const maybePath =
              !isInplace() && maybeTo
                ? [maybeTo.state.currDir.path]
                : [FileListState.currentItem(from.state)]
                    .filter((_) => _ !== undefined)
                    .map((_) => _.name);
            return maybePath.length > 0 ? maybePath[0] : "";
          })(),
          items,
          onAction,
          onCancel: props.onClose,
        })
      : maybeError !== undefined
        ? h(messageBoxComp, {
            title: "Error",
            message: maybeError,
            actions: [MessageBoxAction.OK(props.onClose)],
            style: currTheme.popup.error,
          })
        : showStats
          ? h(copyItemsStats, {
              dispatch: from.dispatch,
              actions: from.actions,
              fromPath: from.state.currDir.path,
              items,
              title: move ? "Move" : "Copy",
              onDone: (total) => {
                setTotal(total);
                setShowStats(false);
              },
              onCancel: props.onClose,
            })
          : showMove && maybeToPath !== undefined
            ? (([path, toPath]) => {
                return h(moveProcessComp, {
                  dispatch: from.dispatch,
                  actions: from.actions,
                  fromPath: from.state.currDir.path,
                  items: !inplace
                    ? items.map((item) => ({ item, toName: item.name }))
                    : items.map((item) => ({ item, toName: toPath })),
                  toPath: !inplace ? toPath : from.state.currDir.path,
                  onTopItem,
                  onDone: onDone(path, toPath),
                });
              })(maybeToPath)
            : maybeTotal !== undefined && maybeToPath !== undefined
              ? ((total, [path, toPath]) => {
                  return h(copyProcessComp, {
                    from,
                    to: !inplace && maybeTo ? maybeTo : from,
                    move,
                    fromPath: from.state.currDir.path,
                    items: !inplace
                      ? items.map((item) => ({ item, toName: item.name }))
                      : items.map((item) => ({ item, toName: toPath })),
                    toPath: !inplace ? toPath : from.state.currDir.path,
                    total,
                    onTopItem,
                    onDone: onDone(path, toPath),
                  });
                })(maybeTotal, maybeToPath)
              : null;
  };

  CopyMoveUiComp.displayName = "CopyMoveUi";

  return CopyMoveUiComp;
}

CopyMoveUi.copyItemsPopup = CopyItemsPopup;
CopyMoveUi.copyItemsStats = CopyItemsStats;
CopyMoveUi.messageBoxComp = MessageBox;
CopyMoveUi.moveProcessComp = MoveProcess;
CopyMoveUi.copyProcessComp = CopyProcess;

/**
 * @param {FileListData} from
 * @param {string} toPath
 * @returns {Promise<boolean>}
 */
async function checkSameDrive(from, toPath) {
  const maybeFromRoot = await from.actions.api.getDriveRoot(
    from.state.currDir.path,
  );
  const maybeToRoot = await from.actions.api.getDriveRoot(toPath);
  return (
    maybeFromRoot !== undefined &&
    maybeToRoot !== undefined &&
    maybeFromRoot === maybeToRoot
  );
}

export default CopyMoveUi;
